src/win.rs

```diff
@@ -6,61 +6,69 @@ use winapi::um::securitybaseapi::*;
 use winapi::um::winnt::*;
 
+use futures::Stream;
 use std::io;
 use std::marker;
 use std::mem;
-use std::ptr;
-use futures::Stream;
-use tokio::prelude::*;
+use std::path::Path;
 use std::pin::Pin;
+use std::ptr;
 use std::task::{Context, Poll};
-use std::path::Path;
-use std::mem::MaybeUninit;
-use tokio::io::PollEvented;
+use std::time::{Duration, Instant};
+use tokio::io::{AsyncRead, AsyncWrite};
 
-type NamedPipe = PollEvented<mio_named_pipes::NamedPipe>;
+use tokio::net::windows::named_pipe;
 
-const PIPE_AVAILABILITY_TIMEOUT: u64 = 5000;
+enum NamedPipe {
+    Server(named_pipe::NamedPipeServer),
+    Client(named_pipe::NamedPipeClient),
+}
+
+const PIPE_AVAILABILITY_TIMEOUT: Duration = Duration::from_secs(5);
 
 /// Endpoint implementation for windows
 pub struct Endpoint {
     path: String,
     security_attributes: SecurityAttributes,
+    created_listener: bool,
 }
 
 impl Endpoint {
     /// Stream of incoming connections
-    pub fn incoming(mut self) -> io::Result<impl Stream<Item = tokio::io::Result<impl AsyncRead + AsyncWrite>> + 'static> {
-        let pipe = self.inner()?;
-        Ok(Incoming {
-            path: self.path.clone(),
-            inner: NamedPipeSupport {
-                path: self.path,
-                pipe,
-                security_attributes: self.security_attributes,
-            },
-        })
+    pub fn incoming(
+        mut self,
+    ) -> io::Result<impl Stream<Item = io::Result<impl AsyncRead + AsyncWrite>> + 'static> {
+        let pipe = self.create_listener()?;
+
+        let stream =
+            futures::stream::try_unfold((pipe, self), |(listener, mut endpoint)| async move {
+                let () = listener.connect().await?;
+
+                let new_listener = endpoint.create_listener()?;
+
+                let conn = Connection::wrap(NamedPipe::Server(listener));
+
+                Ok(Some((conn, (new_listener, endpoint))))
+            });
+
+        Ok(stream)
     }
 
-    /// Inner platform-dependant state of the endpoint
-    fn inner(&mut self) -> io::Result<NamedPipe> {
-        use miow::pipe::NamedPipeBuilder;
-        use std::os::windows::io::*;
-
-        let raw_handle = unsafe {
-            NamedPipeBuilder::new(&self.path)
-                .first(true)
-                .inbound(true)
-                .accept_remote(false)
-                .outbound(true)
-                .out_buffer_size(65536)
-                .in_buffer_size(65536)
-                .with_security_attributes(self.security_attributes.as_ptr())?
-                .into_raw_handle()
-        };
+    fn create_listener(&mut self) -> io::Result<named_pipe::NamedPipeServer> {
+        let server = unsafe {
+            named_pipe::ServerOptions::new()
+                .first_pipe_instance(!self.created_listener)
+                .reject_remote_clients(true)
+                .access_inbound(true)
+                .access_outbound(true)
+                .create_with_security_attributes_raw(
```

Comment:
Is there buffer configuration too?

Reply:
It is - the old values are also `65536` by default but it's probably a good idea to explicitly state that here as well

Reply:
Done in https://github.com/NikVolf/parity-tokio-ipc/pull/32/commits/5a930fc245adbb9c8874cfd6510170293a6a6ce7